File: dart/game-of-life/lib/game_of_life.dart
typedef Matrix = List<List<int>>;

class GameOfLife {
  Matrix _mtx;

  GameOfLife(this._mtx);

  // TODO: return a clone or a frozen version.
  Matrix matrix() => _mtx;

  Matrix tick() {
    Matrix next = _newMatrix();

    for (var row = 0; row < _mtx.length; row++) {
      for (var col = 0; col < _mtx[row].length; col++) {
        next[row][col] = switch (_countNeighbours(row, col)) {
          3 => 1,               // becomes live
          2 => _mtx[row][col],  // stays live if live
          _ => 0,               // dies
        };
      }
    }

    _mtx = next;
    return next;
  }

  Matrix _newMatrix() =>
      List.generate(_mtx.length, (i) => List.filled(_mtx[i].length, 0));

  int _countNeighbours(int row, int col) {
    var count = 0;

    for (var delta in [[-1,-1], [-1,0], [-1,1], [0,-1], [0,1], [1,-1], [1,0], [1,1]]) {
      var dr = row + delta[0];
      var dc = col + delta[1];
      if (0 <= dr && dr < _mtx.length &&
          0 <= dc && dc < _mtx[row].length) { count += _mtx[dr][dc]; }
    }

    return count;
  }
}

File: javascript/game-of-life/game-of-life.js
const clone = (matrix) => JSON.parse(JSON.stringify(matrix));

export class GameOfLife {
  constructor(matrix) {
    this.matrix = clone(matrix);
    this.nrows = matrix.length;
    if (this.nrows) this.ncols = matrix[0].length;
  }

  tick() {
    const next = clone(this.matrix);
    for (let row = 0; row < this.nrows; row++) {
      for (let col = 0; col < this.ncols; col++) {
        const ns = this.countNeighbours(row, col);
        switch (ns) {
          case 3:  next[row][col] = 1; break;
          case 2:  next[row][col] = this.matrix[row][col]; break;
          default: next[row][col] = 0;
        }
      }
    }
    this.matrix = next;
  }

  state() {
    return this.matrix;
  }

  countNeighbours(row, col) {
    let n = 0;
    for (let rr = row - 1; rr <= row + 1; rr += 1)
      if (0 <= rr && rr < this.ncols)
        for (let cc = col - 1; cc <= col + 1; cc += 1)
          if ((0 <= cc && cc < this.nrows) && (row !== rr || col !== cc))
            n += this.matrix[rr][cc];
    return n;
  }
}
